src/javascript/crypto/e2e/openpgp/pgpmime/mimenode.js

```diff
@@ -0,0 +1,198 @@
+
+goog.require('goog.array');
+goog.require('goog.asserts');
+goog.require('goog.crypt.base64');
+goog.require('goog.object');
+goog.require('goog.string');
+
+goog.scope(function() {
+var pgpmime = e2e.openpgp.pgpmime;
+var constants = pgpmime.Constants;
+
+
+
+/**
+ * Constructor for a MIME tree node.
+ * @param {{
+ *   multipart: boolean,
+ *   optionalHeaders: (Array<!e2e.openpgp.pgpmime.types.Header>|undefined)
+ * }} options Options to initialize for the node.
+ * @param {e2e.openpgp.pgpmime.MimeNode=} opt_parent The parent node.
+ * @constructor
+ */
+e2e.openpgp.pgpmime.MimeNode = function(options, opt_parent) {
+  this.parent = opt_parent || this;
+  this.multipart_ = options.multipart;
+
+  this.children_ = [];
+  this.header_ = /** @type {e2e.openpgp.pgpmime.types.Header} */ ({});
+  this.content_ = null;
+  this.boundary_ = '';
+
+  // set headers
+  if (goog.isDefAndNotNull(options.optionalHeaders)) {
+    goog.array.forEach(options.optionalHeaders, goog.bind(function(header) {
+      if (goog.isDefAndNotNull(header.name) &&
+          goog.isDefAndNotNull(header.value)) {
+        if (goog.isDefAndNotNull(header.params)) {
+          this.setHeader_(header.name, header.value, header.params);
+        } else {
+          this.setHeader_(header.name, header.value);
+        }
+      }
+    }, this));
+  }
+
+  this.setBoundary_();
+};
+
+
+/**
+ * Sets the MIME message boundary for a node.
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setBoundary_ = function() {
+  // TODO: Strictly ensure that the boundary value doesn't coincide with
+  //   any string in the email content and headers.
+  this.boundary_ = '---' + goog.string.getRandomString() +
+      Math.floor(Date.now() / 1000).toString();
+};
+
+
+/**
+ * Adds a child to a MIME node.
+ * @param {{
+ *   multipart: boolean,
+ *   optionalHeaders: (Array<!e2e.openpgp.pgpmime.types.Header>|undefined)
+ * }} options Options to initialize for the node.
+ * @return {e2e.openpgp.pgpmime.MimeNode}
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.addChild = function(options) {
+  var node = new e2e.openpgp.pgpmime.MimeNode(options, this);
+  this.children_.push(node);
+  return node;
+};
+
+
+/**
+ * Sets a MIME header.
+ * @param {string} key Name of the header.
+ * @param {string} value Value of the header, possibly including parameters.
+ * @param {Object.<string, string>=} opt_params Optional dict of additional
+ *   parameters
+ * @private
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setHeader_ =
+    function(key, value, opt_params) {
+  var headerValue = e2e.openpgp.pgpmime.Utils.parseHeaderValue(value);
+  if (opt_params) {
+    headerValue.params = headerValue.params || {};
+    goog.object.extend(headerValue.params, opt_params);
+  }
+  goog.object.set(this.header_, key, headerValue);
+};
+
+
+/**
+ * Sets the content.
+ * @param {(string|!e2e.ByteArray)} content The content to set
+ */
+e2e.openpgp.pgpmime.MimeNode.prototype.setContent = function(content) {
```

Comment:
`ByteArray` content will fail in buildContent (it's always `typeof string` or `goog.assertString`). In the test the BINARY_CONTENT is a string, and not a byte array.

In general it's better to just accept one type, `ByteArray` being the more generic. we have conversion functions in `e2e` namespace.